text/0022-response-context.md

```diff
@@ -0,0 +1,92 @@
+* Start Date: 2022-10-03
+* RFC Type: feature
+* RFC PR: https://github.com/getsentry/rfcs/blob/rfc/response-context/text/0022-response-context.md
+
+# Summary
+
+Add `Response` interface that contains information on a HTTP response related to the event.
+
+# Motivation
+
+The [Request](https://develop.sentry.dev/sdk/event-payloads/request/) interface contains information on a HTTP request related to the event. However, there is no interface that contains information on a HTTP response related to the event. This RFC proposes a new interface called `Response` that contains information on a HTTP response related to the event.
+
+The `Request` interface has a few limitations:
+* Does not accept arbitrary fields, so unknown fields are dropped during ingestion such as `status_code`.
+* Response `headers` are not dropped during data scrubbing, See [issue](https://github.com/getsentry/relay/issues/1501).
+* Adding `Response` data in the `Request` interface is semantically wrong anyways.
+
+This is necessary to implement the `Failed HTTP Client requests automatically result in Events` feature that is described in this [DACI](https://www.notion.so/sentry/Failed-HTTP-Client-requests-automatically-result-in-Events-f6c21d2a58ce4f2c889a823fd1da0044).
+
+# Background
+
+Since the `Response` metadata is PII sensitive, it should be properly mapped, documented and scrubbed.
+
+# Supporting Data
+
+See [Supporting data](https://www.notion.so/sentry/Failed-HTTP-Client-requests-automatically-result-in-Events-f6c21d2a58ce4f2c889a823fd1da0044#0ca951d5216742dbaab02f5fd33b8fb5) section in the [DACI](https://www.notion.so/sentry/Failed-HTTP-Client-requests-automatically-result-in-Events-f6c21d2a58ce4f2c889a823fd1da0044).
+
+# Options Considered
+
+## Option 1 (Preferred)
+
+Adding a `Response` interface directly in the Event payload.
+
+```json
+{
+  "response": {
+    "method": "POST",
+    "url": "http://absolute.uri/foo",
+    "query_string": "query=foobar&page=2",
+    "data": { // arbitrary fields
+      "foo": "bar"
+    },
+    "cookies": "PHPSESSID=298zf09hf012fh2; csrftoken=u32t4o3tb3gg43; _gat=1;",
+    "headers": {
+      "content-type": "text/html"
+    },
+    "env": {
+      "REMOTE_ADDR": "192.168.0.1"
+    },
+    "status_code": 500,
+    "status_message": "foo bar", // is that useful?
+    "is_redirect": false,
+    "response_body_size": 1000, // bytes
+  }
+}
+```
+
+The `data` field is semantically different than the `Request#data`. The `Request#data` is the data that was sent to the server, while the `Response#data` is just arbitrary data that is attached to the response. The `Response#data` is not used for anything in the SDKs, but it is useful for the user to attach arbitrary data to the response.
+We could rename to `other` or something else to avoid confusion.
```

Comment:
It's not mandatory, we can remove it if it's confusing, I was just thinking that in case we wanna add something more to the `Response` protocol, it'd require adding to Relay, etc, with an arbitrary data field, I can just add something in there easily with back compatibility.

Reply:
I'd not add the `body` field because there are problems that come with it, and we agreed that the `body` would not make it into this feature yet.
Ideally, we'd add the `request` and `response` bodies via `attachments` to not exceed the max size of events. (1MB currently).
Also, we'd need to scrub (PII) very carefully.

Reply:
> I was just thinking that in case we wanna add something more to the Response protocol, it'd require adding to Relay
I suspect that we can also make Relay accept and pass on any unknown fields in the response context (top level).

Reply:
That would be awesome if possible.

Reply:
TLDR: Either we have a data bag that accepts arbitrary key/values or the response protocol should retain unknown (high level) fields.

Reply:
If the purpose of the databag is to blindly forward all possible values into the database then we might as well accept arbitrary additional fields. However I would prefer if we did neither of those and type out fields we want to send properly, for documentation purposes.

Reply:
it's one or the other, not both.

> arbitrary and retained fields (either this or data)

Reply:
I'm fine not having both, but its a problem for back compatibility when adding new fields (older self-hosted versions will drop fields), and its also not that flexible since the user always know more than the SDK and can add more context that helps identify the issue.
I'm fine either way, either the databag or arbitrary fields are retained, any preference?

Reply:
@untitaker we do have `retain = "true"` on all other contexts as far as I can see, so I would just do the same here.

https://github.com/getsentry/relay/blob/eace81d7f4ad8ecf3a94e76738e9e9bce9639116/relay-general/src/protocol/contexts.rs#L580-L582

@marandaneto I definitely prefer arbitrary fields over a dedicated databag. But the goal should be to add any new field to the schema eventually.

Reply:
Sounds good, I will remove the databag from the RFC.